System.Security.Cryptography/HMACSHA256.cs

```diff
@@ -0,0 +1,169 @@
+        /// </remarks>
+        public HMACSHA256(byte[] key)
+        {
+            _keyValue = key ?? throw new ArgumentNullException();
+        }
+
+        /// <summary>
+        /// Computes the hash value for the specified byte array.
+        /// </summary>
+        /// <param name="buffer">The input to compute the hash code for.</param>
+        /// <returns>The computed hash code.</returns>
+        /// <exception cref="ObjectDisposedException">The object has already been disposed.</exception>
+        /// <exception cref="ArgumentNullException"><paramref name="buffer"/> is <see langword="null"/>.</exception>
+        public byte[] ComputeHash(byte[] buffer)
+        {
+            if (_disposed)
+            {
+                throw new ObjectDisposedException();
+            }
+
+            if (buffer == null)
+            {
+                throw new ArgumentNullException();
+            }
+
+            _hashValue = HashCore(_keyValue, buffer);
+
+            return (byte[])_hashValue.Clone();
+        }
+
+        /// <summary>
+        /// Computes the HMAC of data using the SHA256 algorithm.
+        /// </summary>
+        /// <param name="key">The HMAC key.</param>
+        /// <param name="source">The data to HMAC.</param>\
+        /// <returns>The HMAC of the data.</returns>
+        /// <exception cref="ArgumentNullException">
+        /// <paramref name="key" /> or <paramref name="source" /> is <see langword="null" />.
+        /// </exception>
+        public static byte[] HashData(
+            byte[] key,
+            byte[] source)
+        {
+            // key and source are checked for null by HashCore
+
+            return HashCore(key, source);
+        }
+
+        /// <inheritdoc/>
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        /// <inheritdoc/>
+        protected void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                if (_keyValue != null)
+                {
+                    Array.Clear(_keyValue, 0, _keyValue.Length);
+                }
+
+                _keyValue = null;
+
+                // Although we don't have any resources to dispose at this level,
+                // we need to continue to throw ObjectDisposedExceptions from CalculateHash
+                // for compatibility with the .NET Framework.
+                _disposed = true;
+            }
+
+            return;
+        }
+
+        [MethodImpl(MethodImplOptions.InternalCall)]
+        extern private static byte[] HashCore(byte[] key, byte[] source);
```

Comment:
Is this always present in all the firmware? Otherwise we should provide as well a pure managed version for modem implementation where we want to be able to run this on any system, not just those have System.Net enabled.

Reply:
System.Net is not a requirement to have this.
Just like any other assembly the availability on the fw image is dependent on a build option.
We can bring in this independent of it.
If the target doesn't have support for it, VS won't deploy.

Reply:
but then, we are loosing the capability to have this done on a pure managed way. I would prefer to have then a native pure software implementation when it's not implemented with some native platform functions.
And somehow have those few API always available on the native side. So regardless, we're sure we can count on them.

Reply:
we just added modem support with purely managed classes. So then, we would have to provide as well, like it is today and only for the purely managed the HMACCSHA256 (and SHA256) as purely managed. So we will have this purely managed implementation.
And so, then, ok to have this one not on all devices and on build. The path then should be clear:
-> wifi/ethnet => the Azure lib + all this new nuget
-> modem => managed Azure lib only (it will come with managed crypto)

I think I'm fine with this! (not CRC32 is a different discussion ;-))

Reply:
I actually see 3 paths here... each with benefits and pitfalls...

1) in the managed class, if an exception is thrown (because the feature is not available) call a managed class.

2) implement the "native" software implementation.

3) throw a warning to "suggest" using an independent lib that includes it.

Reply:
Adding a forth...

Have 2 nugets available, one purely managed and one using the native implementation... We kind of do that for some other libs.

Reply:
On a project referencing Sys.Sec.Cryptography, if this is not available in the target device, the deployment won't even happen, so 1. above is technically impossible.

All these cryptography implementations are supplied through Mbed TLS. They are all software based. On some platforms there are hardware accelerators for some algorithms, but this is a just a detail on the implementation.

I'm convinced that enabling cryptography for all targets it's a mistake because those that do not have network enabled (and please keep in mind that this is independent of it, just mentioning it for the sake of this conversation) are meant to be slim and have no use for it. This is exposed as a build option and one can always build this if needed.

We can also split the distribution of this library by the various features, but maybe that would be overengineering this a bit too much and getting into a granularity level that's not reasonable...